src/JetMETAnalyzer.cc

```diff
@@ -0,0 +1,517 @@
+         auto part = jet.daughterPtr(j);
+         if (!( part.isAvailable() && part.isNonnull()) ){
+             continue;	
+         }
+
+         if (fabs(part->charge()) > 0) {
+             nCh_tmp++;
+         }
+         else if(fabs(part->charge())==0){
+             nNeutrals_tmp++;
+         }
+
+         float deta = part->eta() - etaJet;
+         float dphi = 2*atan(tan((part->phi()-phiJet)/2));
+         float dR = sqrt(deta*deta + dphi*dphi);
+         float weight = part->pt();
+         float weight2 = weight * weight;
+         sumWdR2      +=weight2*dR*dR;
+         sumW         += weight;
+         sumW2        += weight2;
+
+         if (dR < 0.1) {
+             sum_ring0 += weight;
+         }
+         else if (dR >= 0.1 && dR < 0.2) {
+             sum_ring1 += weight;
+         }
+         else if (dR >= 0.2 && dR < 0.3) {
+             sum_ring2 += weight;
+         }
+         else if (dR >= 0.3 && dR < 0.4){
+             sum_ring3 += weight;
+         }
+         else if (dR >= 0.4 && dR < 0.5){
+             sum_ring4 += weight;
+         }
+         else if (dR >= 0.5 && dR < 0.6){
+             sum_ring5 += weight;
+         }
+         else if (dR >= 0.6 && dR < 0.7){
+             sum_ring6 += weight;
+         }
+         else if (dR >= 0.7 && dR < 0.8){
+             sum_ring7 += weight;
+         }
+         else{
+             sum_ring8 += weight;
+         }
+
+         reco::CandidatePtr pfJetConstituent = jet.sourceCandidatePtr(j);
+         const reco::Candidate* icand = pfJetConstituent.get();
+         const pat::PackedCandidate* lPack = dynamic_cast<const pat::PackedCandidate *>( icand );
+         if (lPack) {
+
+             if(fabs(lPack->charge()) > 0){
+
+                 if(lPack->pt() > pTMax){
+                     pTMax = lPack->pt();
+                     dZ2 = lPack->dz();
+                 }
+
+                 float tkpt = lPack->pt();
+                 sumTkPt += tkpt;
+                 bool inVtx0 = (lPack->fromPV()==3);
+                 bool inAnyOther = (lPack->fromPV()==0);
+                 double dZ0 = lPack->dz();
+                 double dZ_tmp = dZ0;
+
+                 for(reco::VertexCollection::const_iterator  vi=vtx->begin(); vi!=vtx->end(); ++vi ) {
+                     const reco::Vertex & iv = *vi;
+                     if( iv.isFake() ) { continue; }
+                     if(fabs(lPack->dz(iv.position()))<fabs(dZ_tmp)){dZ_tmp=lPack->dz(iv.position());}
+                 }
+
+                 if( inVtx0 && ! inAnyOther ) {
+                     betaClassic_tmp += tkpt;
+                 }
+                 else if( ! inVtx0 && inAnyOther ) {
+                     betaStarClassic_tmp += tkpt;
+                 }
+                 if( fabs(dZ0) < 0.2 ) {
+                     beta_tmp += tkpt;
+                 }
+                 else if( fabs(dZ_tmp) < 0.2 ) {
+                     betaStar_tmp += tkpt;
+                 }
+             }
+         }
+     }
```

Comment:
Yep sure.

I just used the previous code without doing any modifications, but I'll clean it up and update this PR.

Thanks for the detailed review :)